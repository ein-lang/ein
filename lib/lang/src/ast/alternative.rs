use super::expression::Expression;
use crate::types::Type;

#[derive(Clone, Debug, PartialEq)]
pub struct Alternative {
    type_: Type,
    expression: Expression,
}

impl Alternative {
    pub fn new(type_: impl Into<Type>, expression: impl Into<Expression>) -> Self {
        Self {
            type_: type_.into(),
            expression: expression.into(),
        }
    }

    pub fn type_(&self) -> &Type {
        &self.type_
    }

    pub fn expression(&self) -> &Expression {
        &self.expression
    }

    pub fn transform_expressions<E>(
        &self,
        transform: &mut impl FnMut(&Expression) -> Result<Expression, E>,
    ) -> Result<Self, E> {
        Ok(Self::new(
            self.type_.clone(),
            self.expression.transform_expressions(transform)?,
        ))
    }

    pub fn transform_types<E>(
        &self,
        transform: &mut impl FnMut(&Type) -> Result<Type, E>,
    ) -> Result<Self, E> {
        Ok(Self::new(
            self.type_.transform_types(transform)?,
            self.expression.transform_types(transform)?,
        ))
    }
}
