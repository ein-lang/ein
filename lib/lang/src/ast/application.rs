use super::expression::Expression;
use crate::{
    debug::SourceInformation,
    types::{self, Type},
};
use std::sync::Arc;

#[derive(Clone, Debug, PartialEq)]
pub struct Application {
    type_: Type,
    function: Arc<Expression>,
    argument: Arc<Expression>,
    source_information: Arc<SourceInformation>,
}

impl Application {
    pub fn new(
        function: impl Into<Expression>,
        argument: impl Into<Expression>,
        source_information: impl Into<Arc<SourceInformation>>,
    ) -> Self {
        let source_information = source_information.into();

        Self::with_type(
            types::Unknown::new(source_information.clone()),
            function,
            argument,
            source_information,
        )
    }

    pub fn with_type(
        type_: impl Into<Type>,
        function: impl Into<Expression>,
        argument: impl Into<Expression>,
        source_information: impl Into<Arc<SourceInformation>>,
    ) -> Self {
        Self {
            type_: type_.into(),
            function: Arc::new(function.into()),
            argument: Arc::new(argument.into()),
            source_information: source_information.into(),
        }
    }

    pub fn type_(&self) -> &Type {
        &self.type_
    }

    pub fn function(&self) -> &Expression {
        &self.function
    }

    pub fn argument(&self) -> &Expression {
        &self.argument
    }

    pub fn source_information(&self) -> &Arc<SourceInformation> {
        &self.source_information
    }

    pub fn transform_expressions<E>(
        &self,
        transform: &mut impl FnMut(&Expression) -> Result<Expression, E>,
    ) -> Result<Self, E> {
        Ok(Self::with_type(
            self.type_.clone(),
            self.function.transform_expressions(transform)?,
            self.argument.transform_expressions(transform)?,
            self.source_information.clone(),
        ))
    }

    pub fn transform_types<E>(
        &self,
        transform: &mut impl FnMut(&Type) -> Result<Type, E>,
    ) -> Result<Self, E> {
        Ok(Self::with_type(
            self.type_.transform_types(transform)?,
            self.function.transform_types(transform)?,
            self.argument.transform_types(transform)?,
            self.source_information.clone(),
        ))
    }
}
