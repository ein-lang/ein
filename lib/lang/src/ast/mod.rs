mod alternative;
mod application;
mod arithmetic_operation;
mod boolean;
mod boolean_operation;
mod case;
mod definition;
mod equality_operation;
mod export;
mod expression;
mod function_definition;
mod if_;
mod import;
mod let_;
mod list;
mod list_case;
mod list_element;
mod module;
mod module_interface;
mod none;
mod number;
mod operation;
mod order_operation;
mod record_construction;
mod record_element_operation;
mod record_update;
mod string;
mod type_coercion;
mod type_definition;
mod unresolved_import;
mod unresolved_module;
mod variable;
mod variable_definition;

pub use alternative::*;
pub use application::*;
pub use arithmetic_operation::*;
pub use boolean::*;
pub use boolean_operation::*;
pub use case::*;
pub use definition::*;
pub use equality_operation::*;
pub use export::*;
pub use expression::*;
pub use function_definition::*;
pub use if_::*;
pub use import::*;
pub use let_::*;
pub use list::*;
pub use list_case::*;
pub use list_element::*;
pub use module::*;
pub use module_interface::*;
pub use none::*;
pub use number::*;
pub use operation::*;
pub use order_operation::*;
pub use record_construction::*;
pub use record_element_operation::*;
pub use record_update::*;
pub use string::*;
pub use type_coercion::*;
pub use type_definition::*;
pub use unresolved_import::*;
pub use unresolved_module::*;
pub use variable::*;
pub use variable_definition::*;
