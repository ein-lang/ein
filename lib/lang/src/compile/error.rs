use crate::{ast, debug::*, path::ModulePath, types};
use std::{error::Error, fmt::Display, sync::Arc};

#[derive(Debug, PartialEq)]
pub enum CompileError {
    AnyEqualOperation(Arc<SourceInformation>),
    CaseArgumentTypeInvalid(Arc<SourceInformation>),
    DuplicateNames(Arc<SourceInformation>, Arc<SourceInformation>),
    ExportedNameNotFound {
        name: String,
    },
    FunctionEqualOperation(Arc<SourceInformation>),
    FunctionExpected(Arc<SourceInformation>),
    MainFunctionNotFound(ModulePath),
    RecordElementNotFound {
        record_type: types::Record,
        name: String,
    },
    RecordEqualOperation(Arc<SourceInformation>),
    EirFmmCompile(eir_fmm::CompileError),
    TypeNotFound(types::Reference),
    TypesNotMatched(Arc<SourceInformation>, Arc<SourceInformation>),
    TypeNotInferred(Arc<SourceInformation>),
    VariableNotFound(ast::Variable),
}

impl Display for CompileError {
    fn fmt(&self, formatter: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
        match self {
            Self::AnyEqualOperation(source_information) => write!(
                formatter,
                "cannot compare Any type values\n{}",
                source_information
            ),
            Self::CaseArgumentTypeInvalid(source_information) => write!(
                formatter,
                "invalid argument type of case expression\n{}",
                source_information
            ),
            Self::ExportedNameNotFound { name } => {
                write!(formatter, "exported name \"{}\" not found", name)
            }
            Self::DuplicateNames(one, other) => {
                write!(formatter, "duplicate names\n{}\n{}", one, other)
            }
            Self::FunctionEqualOperation(source_information) => write!(
                formatter,
                "cannot compare functions\n{}",
                source_information
            ),
            Self::FunctionExpected(source_information) => {
                write!(formatter, "function expected\n{}", source_information)
            }
            Self::MainFunctionNotFound(path) => write!(
                formatter,
                "main function not found in main module {}",
                &path
            ),
            Self::RecordElementNotFound { record_type, name } => write!(
                formatter,
                "element \"{}\" not found in record type\n{}",
                &name,
                record_type.source_information()
            ),
            Self::RecordEqualOperation(source_information) => write!(
                formatter,
                "cannot compare records including functions or Any values\n{}",
                source_information
            ),
            Self::EirFmmCompile(error) => {
                write!(formatter, "failed to compile eir to fmm: {:?}", error)
            }
            Self::TypeNotFound(reference) => write!(
                formatter,
                "type \"{}\" not found\n{}",
                reference.name(),
                reference.source_information()
            ),
            Self::TypeNotInferred(source_information) => {
                write!(formatter, "failed to infer type\n{}", source_information)
            }
            Self::TypesNotMatched(lhs_source_information, rhs_source_information) => write!(
                formatter,
                "types not matched\n{}\n{}",
                lhs_source_information, rhs_source_information
            ),
            Self::VariableNotFound(variable) => write!(
                formatter,
                "variable \"{}\" not found\n{}",
                variable.name(),
                variable.source_information()
            ),
        }
    }
}

impl Error for CompileError {}

impl From<eir_fmm::CompileError> for CompileError {
    fn from(error: eir_fmm::CompileError) -> Self {
        Self::EirFmmCompile(error)
    }
}
