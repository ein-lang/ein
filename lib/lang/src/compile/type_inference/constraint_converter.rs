use super::{
    super::{error::CompileError, reference_type_resolver::ReferenceTypeResolver},
    variable_constraint::VariableConstraint,
};
use crate::types::{self, Type};
use std::sync::Arc;

pub struct ConstraintConverter {
    reference_type_resolver: Arc<ReferenceTypeResolver>,
}

impl ConstraintConverter {
    pub fn new(reference_type_resolver: Arc<ReferenceTypeResolver>) -> Arc<Self> {
        Self {
            reference_type_resolver,
        }
        .into()
    }

    pub fn convert(&self, constraint: &VariableConstraint) -> Result<Type, CompileError> {
        Ok(
            if constraint
                .lower_types()
                .iter()
                .map(|type_| self.reference_type_resolver.is_any(type_))
                .collect::<Result<Vec<bool>, _>>()?
                .into_iter()
                .any(|ok| ok)
            {
                types::Any::new(constraint.source_information().clone()).into()
            } else if !constraint.lower_types().is_empty() {
                types::Union::new(
                    constraint.lower_types().iter().cloned().collect(),
                    constraint.source_information().clone(),
                )
                .into()
            } else if !constraint.upper_types().is_empty() {
                // TODO Calculate intersection types from upper types?
                constraint.upper_types().iter().next().unwrap().clone()
            } else {
                types::Unknown::new(constraint.source_information().clone()).into()
            },
        )
    }
}
