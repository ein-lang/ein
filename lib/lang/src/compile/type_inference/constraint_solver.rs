use super::{
    super::{error::CompileError, reference_type_resolver::ReferenceTypeResolver},
    constraint_converter::ConstraintConverter,
    subsumption_set::SubsumptionSet,
    variable_constraint_set::VariableConstraintSet,
};
use crate::types::Type;
use std::{collections::HashMap, sync::Arc};

pub struct ConstraintSolver {
    constraint_converter: Arc<ConstraintConverter>,
    reference_type_resolver: Arc<ReferenceTypeResolver>,
}

impl ConstraintSolver {
    pub fn new(
        constraint_converter: Arc<ConstraintConverter>,
        reference_type_resolver: Arc<ReferenceTypeResolver>,
    ) -> Arc<Self> {
        Self {
            constraint_converter,
            reference_type_resolver,
        }
        .into()
    }

    pub fn solve(
        &self,
        mut solved_subsumption_set: SubsumptionSet,
        checked_subsumption_set: &mut SubsumptionSet,
    ) -> Result<HashMap<usize, Type>, CompileError> {
        let mut constraint_set = VariableConstraintSet::new();

        while let Some(subsumption) = solved_subsumption_set.remove() {
            match subsumption {
                (lower, Type::Variable(variable)) => {
                    if let Type::Variable(lower) = &lower {
                        if variable.id() == lower.id() {
                            continue;
                        }
                    }

                    for type_ in constraint_set.get_upper_types(variable.id()) {
                        solved_subsumption_set.add(lower.clone(), type_.clone());
                    }

                    constraint_set.add_lower_type(&variable, &lower);
                }
                (Type::Variable(variable), upper) => {
                    for type_ in constraint_set.get_lower_types(variable.id()) {
                        solved_subsumption_set.add(type_.clone(), upper.clone());
                    }

                    constraint_set.add_upper_type(&variable, &upper);
                }
                (Type::Reference(reference), upper) => solved_subsumption_set.add(
                    self.reference_type_resolver.resolve_reference(&reference)?,
                    upper.clone(),
                ),
                (lower, Type::Reference(reference)) => solved_subsumption_set.add(
                    lower.clone(),
                    self.reference_type_resolver.resolve_reference(&reference)?,
                ),
                (Type::Function(one), Type::Function(other)) => {
                    solved_subsumption_set.add(other.argument().clone(), one.argument().clone());
                    checked_subsumption_set.add(one.argument().clone(), other.argument().clone());
                    solved_subsumption_set.add(one.result().clone(), other.result().clone());
                    checked_subsumption_set.add(other.result().clone(), one.result().clone());
                }
                (Type::List(one), Type::List(other)) => {
                    solved_subsumption_set.add(one.element().clone(), other.element().clone());
                }
                subsumption => checked_subsumption_set.add(subsumption.0, subsumption.1),
            }
        }

        self.convert_to_substitutions(&constraint_set)
    }

    fn convert_to_substitutions(
        &self,
        constraint_set: &VariableConstraintSet,
    ) -> Result<HashMap<usize, Type>, CompileError> {
        let mut substitutions = HashMap::<usize, Type>::new();

        for (id, constraint) in constraint_set.constraints() {
            let constraint_type = self
                .constraint_converter
                .convert(constraint)?
                .substitute_variables(&substitutions);

            for type_ in substitutions.values_mut() {
                *type_ = type_.substitute_variable(*id, &constraint_type);
            }

            substitutions.insert(*id, constraint_type);
        }

        Ok(substitutions)
    }
}
