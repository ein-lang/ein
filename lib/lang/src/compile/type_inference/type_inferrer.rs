use super::{
    super::{
        error::CompileError, reference_type_resolver::ReferenceTypeResolver,
        type_canonicalizer::TypeCanonicalizer, type_equality_checker::TypeEqualityChecker,
    },
    constraint_checker::ConstraintChecker,
    constraint_collector::ConstraintCollector,
    constraint_solver::ConstraintSolver,
    variable_substitutor::VariableSubstitutor,
};
use crate::{
    ast::*,
    types::{self, Type},
};
use std::sync::Arc;

pub struct TypeInferrer {
    reference_type_resolver: Arc<ReferenceTypeResolver>,
    type_equality_checker: Arc<TypeEqualityChecker>,
    type_canonicalizer: Arc<TypeCanonicalizer>,
    constraint_collector: ConstraintCollector,
    constraint_solver: Arc<ConstraintSolver>,
}

impl TypeInferrer {
    pub fn new(
        reference_type_resolver: Arc<ReferenceTypeResolver>,
        type_equality_checker: Arc<TypeEqualityChecker>,
        type_canonicalizer: Arc<TypeCanonicalizer>,
        constraint_collector: ConstraintCollector,
        constraint_solver: Arc<ConstraintSolver>,
    ) -> Self {
        Self {
            reference_type_resolver,
            type_equality_checker,
            type_canonicalizer,
            constraint_collector,
            constraint_solver,
        }
    }

    pub fn infer(self, module: &Module) -> Result<Module, CompileError> {
        let module = module.transform_types(&mut |type_| -> Result<_, CompileError> {
            Ok(match type_ {
                Type::Unknown(unknown) => {
                    types::Variable::new(unknown.source_information().clone()).into()
                }
                _ => type_.clone(),
            })
        })?;

        let (solved_subsumption_set, mut checked_subsumption_set) =
            self.constraint_collector.collect(&module)?;

        let substitutions = self
            .constraint_solver
            .solve(solved_subsumption_set, &mut checked_subsumption_set)?;

        let substitutor = VariableSubstitutor::new(self.type_canonicalizer.clone(), substitutions);

        let checker = ConstraintChecker::new(
            substitutor.clone(),
            self.reference_type_resolver.clone(),
            self.type_equality_checker,
        );

        checker.check(checked_subsumption_set)?;

        module.transform_types(&mut |type_| substitutor.substitute(type_))
    }
}
