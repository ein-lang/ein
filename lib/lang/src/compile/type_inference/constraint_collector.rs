use super::super::error::CompileError;
use super::super::module_environment_creator::ModuleEnvironmentCreator;
use super::super::reference_type_resolver::ReferenceTypeResolver;
use super::subsumption_set::SubsumptionSet;
use crate::ast::*;
use crate::types::{self, Type};
use std::collections::{HashMap, HashSet};
use std::sync::Arc;

pub struct ConstraintCollector {
    reference_type_resolver: Arc<ReferenceTypeResolver>,
    module_environment_creator: Arc<ModuleEnvironmentCreator>,
    solved_subsumption_set: SubsumptionSet,
    checked_subsumption_set: SubsumptionSet,
}

impl ConstraintCollector {
    pub fn new(
        reference_type_resolver: Arc<ReferenceTypeResolver>,
        module_environment_creator: Arc<ModuleEnvironmentCreator>,
    ) -> Self {
        Self {
            reference_type_resolver,
            module_environment_creator,
            solved_subsumption_set: SubsumptionSet::new(),
            checked_subsumption_set: SubsumptionSet::new(),
        }
    }

    pub fn collect(
        mut self,
        module: &Module,
    ) -> Result<(SubsumptionSet, SubsumptionSet), CompileError> {
        let variables = self.module_environment_creator.create(module);

        for definition in module.definitions() {
            match definition {
                Definition::FunctionDefinition(function_definition) => {
                    self.infer_function_definition(function_definition, &variables)?;
                }
                Definition::VariableDefinition(variable_definition) => {
                    self.infer_variable_definition(variable_definition, &variables)?;
                }
            };
        }

        Ok((self.solved_subsumption_set, self.checked_subsumption_set))
    }

    fn infer_function_definition(
        &mut self,
        function_definition: &FunctionDefinition,
        variables: &HashMap<String, Type>,
    ) -> Result<(), CompileError> {
        let source_information = function_definition.source_information();
        let mut variables = variables.clone();
        let mut type_ = function_definition.type_().clone();

        for argument_name in function_definition.arguments() {
            let argument_type: Type = types::Variable::new(source_information.clone()).into();
            let result_type: Type = types::Variable::new(source_information.clone()).into();

            self.solved_subsumption_set.add(
                types::Function::new(
                    argument_type.clone(),
                    result_type.clone(),
                    source_information.clone(),
                ),
                type_,
            );

            variables.insert(argument_name.into(), argument_type);

            type_ = result_type;
        }

        let body_type = self.infer_expression(function_definition.body(), &variables)?;
        self.solved_subsumption_set.add(body_type, type_);

        Ok(())
    }

    fn infer_variable_definition(
        &mut self,
        variable_definition: &VariableDefinition,
        variables: &HashMap<String, Type>,
    ) -> Result<(), CompileError> {
        let type_ = self.infer_expression(variable_definition.body(), &variables)?;

        self.solved_subsumption_set
            .add(type_, variable_definition.type_().clone());

        Ok(())
    }

    fn infer_expression(
        &mut self,
        expression: &Expression,
        variables: &HashMap<String, Type>,
    ) -> Result<Type, CompileError> {
        match expression {
            Expression::Application(application) => {
                let function = self.infer_expression(application.function(), variables)?;
                let argument = self.infer_expression(application.argument(), variables)?;
                let result: Type =
                    types::Variable::new(application.source_information().clone()).into();

                self.solved_subsumption_set.add(
                    function,
                    types::Function::new(
                        argument,
                        result.clone(),
                        application.source_information().clone(),
                    ),
                );

                Ok(result)
            }
            Expression::Boolean(boolean) => {
                Ok(types::Boolean::new(boolean.source_information().clone()).into())
            }
            Expression::Case(case) => {
                let argument = self.infer_expression(case.argument(), variables)?;

                self.checked_subsumption_set
                    .add(argument.clone(), case.type_().clone());
                self.checked_subsumption_set
                    .add(case.type_().clone(), argument);

                let result = types::Variable::new(case.source_information().clone());

                for alternative in case.alternatives() {
                    self.solved_subsumption_set
                        .add(alternative.type_().clone(), case.type_().clone());

                    let mut variables = variables.clone();

                    variables.insert(case.name().into(), alternative.type_().clone());

                    let type_ = self.infer_expression(alternative.expression(), &variables)?;
                    self.solved_subsumption_set.add(type_, result.clone());
                }

                Ok(result.into())
            }
            Expression::If(if_) => {
                let condition = self.infer_expression(if_.condition(), variables)?;
                self.solved_subsumption_set.add(
                    condition,
                    types::Boolean::new(if_.source_information().clone()),
                );

                let then = self.infer_expression(if_.then(), variables)?;
                let else_ = self.infer_expression(if_.else_(), variables)?;
                let result = types::Variable::new(if_.source_information().clone());

                self.solved_subsumption_set.add(then, result.clone());
                self.solved_subsumption_set.add(else_, result.clone());

                Ok(result.into())
            }
            Expression::Let(let_) => {
                let mut variables = variables.clone();

                for definition in let_.definitions() {
                    if let Definition::FunctionDefinition(function_definition) = definition {
                        variables.insert(
                            function_definition.name().into(),
                            function_definition.type_().clone(),
                        );
                    }
                }

                for definition in let_.definitions() {
                    match definition {
                        Definition::FunctionDefinition(function_definition) => {
                            self.infer_function_definition(function_definition, &variables)?;
                        }
                        Definition::VariableDefinition(variable_definition) => {
                            self.infer_variable_definition(variable_definition, &variables)?;

                            variables.insert(
                                variable_definition.name().into(),
                                variable_definition.type_().clone(),
                            );
                        }
                    }
                }

                self.infer_expression(let_.expression(), &variables)
            }
            Expression::List(list) => {
                let element_type = types::Variable::new(list.source_information().clone());
                let list_type =
                    types::List::new(element_type.clone(), list.source_information().clone());

                self.solved_subsumption_set
                    .add(list_type.clone(), list.type_().clone());

                for element in list.elements() {
                    match element {
                        ListElement::Multiple(expression) => {
                            let type_ = self.infer_expression(expression, variables)?;
                            self.solved_subsumption_set.add(type_, list_type.clone())
                        }
                        ListElement::Single(expression) => {
                            let type_ = self.infer_expression(expression, variables)?;
                            self.solved_subsumption_set.add(type_, element_type.clone())
                        }
                    }
                }

                Ok(list_type.into())
            }
            Expression::ListCase(case) => {
                let element = types::Variable::new(case.source_information().clone());
                let list = types::List::new(element.clone(), case.source_information().clone());
                let result = types::Variable::new(case.source_information().clone());

                self.solved_subsumption_set
                    .add(list.clone(), case.type_().clone());

                let argument = self.infer_expression(case.argument(), variables)?;
                self.solved_subsumption_set.add(argument, list.clone());

                let type_ = self.infer_expression(case.empty_alternative(), &variables)?;
                self.solved_subsumption_set.add(type_, result.clone());

                let mut variables = variables.clone();
                variables.insert(case.first_name().into(), element.into());
                variables.insert(case.rest_name().into(), list.into());

                let type_ = self.infer_expression(case.non_empty_alternative(), &variables)?;
                self.solved_subsumption_set.add(type_, result.clone());

                Ok(result.into())
            }
            Expression::None(none) => {
                Ok(types::None::new(none.source_information().clone()).into())
            }
            Expression::Number(number) => {
                Ok(types::Number::new(number.source_information().clone()).into())
            }
            Expression::Operation(operation) => match operation {
                Operation::Boolean(operation) => {
                    let lhs = self.infer_expression(operation.lhs(), variables)?;
                    let rhs = self.infer_expression(operation.rhs(), variables)?;
                    let boolean_type = types::Boolean::new(operation.source_information().clone());

                    self.solved_subsumption_set.add(lhs, boolean_type.clone());
                    self.solved_subsumption_set.add(rhs, boolean_type.clone());

                    Ok(boolean_type.into())
                }
                Operation::Generic(operation) => {
                    let lhs = self.infer_expression(operation.lhs(), variables)?;
                    let rhs = self.infer_expression(operation.rhs(), variables)?;

                    self.solved_subsumption_set
                        .add(lhs.clone(), operation.type_().clone());
                    self.solved_subsumption_set
                        .add(rhs.clone(), operation.type_().clone());

                    Ok(match operation.operator() {
                        Operator::Add
                        | Operator::Subtract
                        | Operator::Multiply
                        | Operator::Divide
                        | Operator::LessThan
                        | Operator::LessThanOrEqual
                        | Operator::GreaterThan
                        | Operator::GreaterThanOrEqual => {
                            let number_type =
                                types::Number::new(operation.source_information().clone());

                            self.solved_subsumption_set.add(lhs, number_type.clone());
                            self.solved_subsumption_set.add(rhs, number_type.clone());

                            match operation.operator() {
                                Operator::Add
                                | Operator::Subtract
                                | Operator::Multiply
                                | Operator::Divide => number_type.into(),
                                _ => types::Boolean::new(operation.source_information().clone())
                                    .into(),
                            }
                        }
                        Operator::Equal | Operator::NotEqual => {
                            types::Boolean::new(operation.source_information().clone()).into()
                        }
                    })
                }
            },
            Expression::RecordConstruction(construction) => {
                let record_type = self
                    .reference_type_resolver
                    .resolve_to_record(construction.type_())?
                    .ok_or_else(|| {
                        CompileError::TypesNotMatched(
                            construction.source_information().clone(),
                            construction.type_().source_information().clone(),
                        )
                    })?;

                if construction.elements().keys().collect::<HashSet<_>>()
                    != record_type.elements().keys().collect()
                {
                    return Err(CompileError::TypesNotMatched(
                        construction.source_information().clone(),
                        record_type.source_information().clone(),
                    ));
                }

                for (key, expression) in construction.elements() {
                    let type_ = self.infer_expression(expression, variables)?;

                    self.solved_subsumption_set
                        .add(type_, record_type.elements()[key].clone());
                }

                Ok(construction.type_().clone())
            }
            Expression::RecordElementOperation(operation) => {
                let record_type = self
                    .reference_type_resolver
                    .resolve_to_record(operation.type_())?
                    .ok_or_else(|| {
                        CompileError::TypesNotMatched(
                            operation.source_information().clone(),
                            operation.type_().source_information().clone(),
                        )
                    })?;

                let argument = self.infer_expression(operation.argument(), variables)?;
                self.solved_subsumption_set
                    .add(argument, operation.type_().clone());

                let mut variables = variables.clone();

                variables.insert(
                    operation.variable().into(),
                    record_type.elements()[operation.key()].clone(),
                );

                self.infer_expression(operation.expression(), &variables)
            }
            Expression::String(string) => {
                Ok(types::EinString::new(string.source_information().clone()).into())
            }
            Expression::Variable(variable) => variables
                .get(variable.name())
                .cloned()
                .ok_or_else(|| CompileError::VariableNotFound(variable.clone())),
            Expression::RecordUpdate(_) | Expression::TypeCoercion(_) => unreachable!(),
        }
    }
}
