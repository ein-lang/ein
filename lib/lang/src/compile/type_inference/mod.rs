mod constraint_checker;
mod constraint_collector;
mod constraint_converter;
mod constraint_solver;
mod subsumption_set;
mod type_inferrer;
mod variable_constraint;
mod variable_constraint_set;
mod variable_substitutor;

use super::builtin_configuration::BuiltinConfiguration;
use super::error::CompileError;
use super::module_environment_creator::ModuleEnvironmentCreator;
use super::reference_type_resolver::ReferenceTypeResolver;
use super::type_canonicalizer::TypeCanonicalizer;
use super::type_equality_checker::TypeEqualityChecker;
use crate::ast::*;
use constraint_collector::ConstraintCollector;
use constraint_converter::ConstraintConverter;
use constraint_solver::ConstraintSolver;
use std::sync::Arc;
use type_inferrer::TypeInferrer;

pub fn infer_types(
    module: &Module,
    builtin_configuration: Arc<BuiltinConfiguration>,
) -> Result<Module, CompileError> {
    let reference_type_resolver = ReferenceTypeResolver::new(&module);
    let type_equality_checker = TypeEqualityChecker::new(reference_type_resolver.clone());
    let type_canonicalizer = TypeCanonicalizer::new(
        reference_type_resolver.clone(),
        type_equality_checker.clone(),
    );
    let constraint_converter = ConstraintConverter::new(reference_type_resolver.clone());
    let module_environment_creator = ModuleEnvironmentCreator::new(builtin_configuration);
    let constraint_collector =
        ConstraintCollector::new(reference_type_resolver.clone(), module_environment_creator);
    let constraint_solver =
        ConstraintSolver::new(constraint_converter, reference_type_resolver.clone());

    TypeInferrer::new(
        reference_type_resolver,
        type_equality_checker,
        type_canonicalizer,
        constraint_collector,
        constraint_solver,
    )
    .infer(module)
}
