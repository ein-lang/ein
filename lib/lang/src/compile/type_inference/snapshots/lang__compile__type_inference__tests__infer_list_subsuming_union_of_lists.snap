---
source: lib/lang/src/compile/type_inference/mod.rs
expression: infer_types(&module)

---
Ok(
    Module {
        path: ModulePath {
            package: Package {
                name: "",
                version: "",
            },
            components: [],
        },
        type_definitions: [],
        definitions: [
            VariableDefinition(
                VariableDefinition {
                    name: "x",
                    body: List(
                        List {
                            type_: List(
                                List {
                                    element: Number(
                                        Number {
                                            source_information: SourceInformation {
                                                source_name: "",
                                                location: Location {
                                                    line_number: 0,
                                                    column_number: 0,
                                                },
                                                line: "",
                                            },
                                        },
                                    ),
                                    source_information: SourceInformation {
                                        source_name: "",
                                        location: Location {
                                            line_number: 0,
                                            column_number: 0,
                                        },
                                        line: "",
                                    },
                                },
                            ),
                            elements: [
                                Single(
                                    Number(
                                        Number {
                                            value: 42.0,
                                            source_information: SourceInformation {
                                                source_name: "",
                                                location: Location {
                                                    line_number: 0,
                                                    column_number: 0,
                                                },
                                                line: "",
                                            },
                                        },
                                    ),
                                ),
                            ],
                            source_information: SourceInformation {
                                source_name: "",
                                location: Location {
                                    line_number: 0,
                                    column_number: 0,
                                },
                                line: "",
                            },
                        },
                    ),
                    type_: Union(
                        Union {
                            types: {
                                List(
                                    List {
                                        element: None(
                                            None {
                                                source_information: SourceInformation {
                                                    source_name: "",
                                                    location: Location {
                                                        line_number: 0,
                                                        column_number: 0,
                                                    },
                                                    line: "",
                                                },
                                            },
                                        ),
                                        source_information: SourceInformation {
                                            source_name: "",
                                            location: Location {
                                                line_number: 0,
                                                column_number: 0,
                                            },
                                            line: "",
                                        },
                                    },
                                ),
                                List(
                                    List {
                                        element: Number(
                                            Number {
                                                source_information: SourceInformation {
                                                    source_name: "",
                                                    location: Location {
                                                        line_number: 0,
                                                        column_number: 0,
                                                    },
                                                    line: "",
                                                },
                                            },
                                        ),
                                        source_information: SourceInformation {
                                            source_name: "",
                                            location: Location {
                                                line_number: 0,
                                                column_number: 0,
                                            },
                                            line: "",
                                        },
                                    },
                                ),
                            },
                            source_information: SourceInformation {
                                source_name: "",
                                location: Location {
                                    line_number: 0,
                                    column_number: 0,
                                },
                                line: "",
                            },
                        },
                    ),
                    source_information: SourceInformation {
                        source_name: "",
                        location: Location {
                            line_number: 0,
                            column_number: 0,
                        },
                        line: "",
                    },
                },
            ),
            VariableDefinition(
                VariableDefinition {
                    name: "y",
                    body: Variable(
                        Variable {
                            name: "x",
                            source_information: SourceInformation {
                                source_name: "",
                                location: Location {
                                    line_number: 0,
                                    column_number: 0,
                                },
                                line: "",
                            },
                        },
                    ),
                    type_: List(
                        List {
                            element: Union(
                                Union {
                                    types: {
                                        None(
                                            None {
                                                source_information: SourceInformation {
                                                    source_name: "",
                                                    location: Location {
                                                        line_number: 0,
                                                        column_number: 0,
                                                    },
                                                    line: "",
                                                },
                                            },
                                        ),
                                        Number(
                                            Number {
                                                source_information: SourceInformation {
                                                    source_name: "",
                                                    location: Location {
                                                        line_number: 0,
                                                        column_number: 0,
                                                    },
                                                    line: "",
                                                },
                                            },
                                        ),
                                    },
                                    source_information: SourceInformation {
                                        source_name: "",
                                        location: Location {
                                            line_number: 0,
                                            column_number: 0,
                                        },
                                        line: "",
                                    },
                                },
                            ),
                            source_information: SourceInformation {
                                source_name: "",
                                location: Location {
                                    line_number: 0,
                                    column_number: 0,
                                },
                                line: "",
                            },
                        },
                    ),
                    source_information: SourceInformation {
                        source_name: "",
                        location: Location {
                            line_number: 0,
                            column_number: 0,
                        },
                        line: "",
                    },
                },
            ),
        ],
        export: Export {
            names: {},
        },
        export_foreign: ExportForeign {
            names: {},
        },
        imports: [],
        import_foreigns: [],
    },
)
